vpr/src/base/vpr_types.h

```diff
@@ -1344,15 +1344,21 @@ struct t_det_routing_arch {
     std::string write_rr_graph_filename;
 };
 
+/*
+ * INC_DIRECTION: wire driver is positioned at the low-coordinate end of the wire.
+ * DEC_DIRECTION: wire_driver is positioned at the high-coordinate end of the wire.
+ * BI_DIRECTION: wire has multiple drivers, so signals can travel either way along the wire
+ * INVALID_DIRECTION: wire does not have a valid direction set
+ */
 enum e_direction : unsigned char {
     INC_DIRECTION = 0,
     DEC_DIRECTION = 1,
     BI_DIRECTION = 2,
-    NO_DIRECTION = 3,
+    INVALID_DIRECTION = 3,
     NUM_DIRECTIONS
 };
 
-constexpr std::array<const char*, NUM_DIRECTIONS> DIRECTION_STRING = {{"INC_DIRECTION", "DEC_DIRECTION", "BI_DIRECTION", "NO_DIRECTION"}};
+constexpr std::array<const char*, NUM_DIRECTIONS> DIRECTION_STRING = {{"INC_DIRECTION", "DEC_DIRECTION", "BI_DIRECTION", "INVALID_DIRECTION"}};
```

Comment:
Can this hold `std::string` instead of `const char*`, so that you can return a const reference to it and don't have to convert it to a std::string every time ? (I presume this is the place where the string conversions come from ?)

Reply:
Good idea; that will remove any (small) performance overhead.

Reply:
@hzeller I am unsure if this is possible. When I try to change the `const char*` to a `std::string`, the compiler gives me the following error: 

`error: the type ‘const std::array<const std::__cxx11::basic_string<char>, 4>’ of ‘constexpr’ variable ‘DIRECTION_STRING’ is not literal
 1361 | constexpr std::array<const std::string, static_cast<int>(e_direction::NUM_DIRECTIONS)> DIRECTION_STRING = {{"INC_DIRECTION", "DEC_DIRECTION", "BI_DIRECTION", "NO_DIRECTION"}};`

Also, this [answer](https://stackoverflow.com/questions/27123306/is-it-possible-to-use-stdstring-in-a-constexpr) on stackoverflow seems to say it is not possible to do this. Perhaps I am missing something, though. Do you have any suggestions here?

Reply:
You're right; from stackoverflow it looks like this isn't doable until C++20.  I wouldn't do it in that case; the performance overhead is minimal and forcing people to use a C++20 standard compiler is probably not a good idea for a few years (we'll get a bunch of build error complaints).

Reply:
Does this array have to be public in the header ? This looks like the original problem exposing such thing in the header.
Usually, I'd just make this a `static std::string foo[] = {"foo", "bar", "baz"};` somewhere in the compilation unit and provide a function to convert the int to the string.

Alternatively as quick hack, the to string function could be to put it in the header as inline function:
```
enum class Dir { FOO, BAR, BAZ };
inline const std::string& direction_string(Dir x) {
  static std::string foo[] = {"foo", "bar", "baz"};
  return foo[(int)x];
}
```

Reply:
I don't think there's any reason it has to be in the header; it would be better to keep it in the .cpp file as you suggest.